### PR TITLE
Warnings and Errors for Type States

### LINKED ISSUE #284: Add semantic/type checking for type states
- Wrong annotations for resulting type
- Call functions without necessary state requirements
- Become after return is accepted
- Become can only use state enums for particular contract
- Attempt call used unnecessarily or needed.

### PR DESCRIPTION
**Pull request by DJRHails**
Fixes #284

# Implementation Overview
Best to look at the test file to see the semantic tests added

- [x] Check invalid characters in state definition
- [x] State has to be specified if contract is stateful
- [x] Check for attempting to become invalid states
- [x] Warning if there are multiple becomes
- [x] Check for code after return/become
- [x] Check for invalid state specifications in restriction group
- [x] Call functions without necessary state requirements
- [x] Become after return is accepted
- [x] Become can only use state enums for particular contract
- [x] Incorrect ordering for become / return
- [x] Becomes are mutating expressions
- [x] Functions with becomes must be mutating

# Notes

# Checklist

- [x] Remove all TODOs, debug prints, whitespace changes
- [x] Include tests
- [x] Add `review wanted` label, remove "WIP" in PR title

### CODE REVIEW

File: Sources/lite/main.swift [hunk 1]
```diff
@@ -58,7 +58,7 @@ func runBehaviorTests() -> Bool {
                                pathExtensions: ["js"],
                                testDirPath: "Tests/BehaviorTests",
                                testLinePrefix: "//",
-                               parallelismLevel: .none,
```
Comment (bearbin) [inline Sources/lite/main.swift]: Was this included by mistake? I thought it worked worse when we tried it at higher parallelism levels.

Comment (DJRHails) [reply Sources/lite/main.swift]: Indeed!

File: Tests/SemanticTests/type_states.flint [hunk 2]
```diff
@@ -1,29 +1,30 @@
 // RUN: %flintc %s --verify
 
-contract TypeStatesInvalid ($state$, ss, TypeStatesInvalid) {}
+contract TypeStatesInvalid ($state$, ss, TypeStatesInvalid) {} // expected-error {{Diallow $ in this identifier too!}}
```
Comment (bearbin) [inline Tests/SemanticTests/type_states.flint]: Is this meant to be a TODO? It so, it shouldn't be too hard to fix.

**Comment by SusanEisenbach** [comment]
I thought we decided return, become.

--
Professor Susan Eisenbach
Department of Computing
Imperial College London

On 2 Aug 2018, at 08:48, Alexander Harkness <notifications@github.com<mailto:notifications@github.com>> wrote:


@bearbin commented on this pull request.

I'm not sure whether the warnings for unexecutable code should be warnings of if errors would be more suitable - if the code is not executable it should not be there to confuse the programmer.

Also, can you remind me which way return/become are supposed to be written:

return x
become y


or

become x
return y


I thought that become would always come after return as return could call a function with state requirements.

________________________________

In Tests/SemanticTests/type_states.flint<https://github.com/flintlang/flint/pull/283#discussion_r207276802>:

> @@ -1,29 +1,30 @@
 // RUN: %flintc %s --verify

-contract TypeStatesInvalid ($state$, ss, TypeStatesInvalid) {}
+contract TypeStatesInvalid ($state$, ss, TypeStatesInvalid) {} // expected-error {{Diallow $ in this identifier too!}}


Is this meant to be a TODO? It so, it shouldn't be too hard to fix.

**Comment by bearbin** [comment]
Yes, that's the way we're going to be doing it. 

Daniel is going to add an error for wrong ordering.

File: Sources/SemanticAnalyzer/SemanticAnalyzer.swift [hunk 3]
```diff
@@ -199,25 +202,36 @@ public struct SemanticAnalyzer: ASTPass {
...
 
-        // Emit a warning if there is code after a return statement.
-        diagnostics.append(.codeAfterReturn(nextStatement))
-      }
-    } else {
-      if let resultType = functionDeclaration.resultType {
-        // Emit an error if a non-void function doesn't have a return statement.
-        diagnostics.append(.missingReturnInNonVoidFunction(closeBraceToken: functionDeclaration.closeBraceToken, resultType: resultType))
-      }
+    let remaingNonEndingStatements = remaining.filter({!$0.isEnding})
```
Comment (franklinsch) [inline Sources/SemanticAnalyzer/SemanticAnalyzer.swift]: Small typo

**Pull request merged**